### ISSUE TITLE
Can we add a stack-size sniffer to our prolog?

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#2809]
Right now, if your code uses more stack space than available (e.g. 80k on a 64k stack), you're probably just going to segfault at runtime. Could we at least add some stack-sniffing code to our prolog that would error-fail in a controlled way, with a useful message?

**Comment by dsharletg** [comment]
This seems difficult to do:

- We can't know reliably how much stack the pipeline wants. We can get a lower bound, but the upper bound is hard to know. So even if we add an assert, it will be fuzzy. We'll either get false positives or false negatives, depending on how conservatively we estimate our stack usage.

- In parallel loops, we can't guarantee anything about what thread we're running on. The master thread participates in the thread pool, so we might need to dynamically grab the stack size on every parallel loop iteration. This will add overhead.

- It's also tricky to do this because the stack gets allocated before any of our code (including this hypothetical check) runs. So we might need to add an extra function call to do the check, then call the real pipeline (with the stack allocation inside it), and require the compiler to *not* inline these calls. For the main pipeline, the overhead of one function call probably doesn't matter, but it would be not-good to do this for every iteration of a parallel loop.

**Comment by zvookin** [comment]
Someone should look at the llvm `stack-probe` function attribute. Typically the way this is done is as follows:

1) Stacks are allocated with a certain number of guard pages at the furthest away from initial use addresses. The number of guard pages determines the larges stack frame one can allocate without an explicit probe. The guard pages are mapped unwritable and likely unreadable. (This requires some OS interaction which may or may not be easy to get into place. E.g. I'm not sure we could do this without fairly platform specific code to make new threads in our thread pool. It certainly won't be the case for a random thread in any old process.)

2) Functions with stack frames smaller than the guard limit just do the normal thing. If they end up causing a stack overflow, the processor faults and the runtime can determine that it is a stack overflow based on the faulting address being in a stack guard page. The function is required to do at least one memory access into its stack frame if it allocates one. (Certainly other bugs can present as faults against the guard pages. Note we aren't trying to recover from stack overflows so much as to report them semi-accurately.)

3) Routines that use a stack frame larger than the guard region size have to call a stack probe routine, thus incurring a bit more overhead. (The stack prober can either do a range check using some thread local that gives the stack bounds or it can try and read from addresses down the stack at increments of the guard region size until it has validated the size of frame the routine needs.)

Obviously figuring out how to do this with the existing llvm support will require a bit of work, but I believe it has functionality to help.

Whether it is worth doing is also open to discussion. As much as it would help productivity in reducing debugging time, likely the immediate response will be "Why is Halide using the stack in the first place?" So my first thought here is we should just add a flag to turn off stack allocation for use inside Google and figure out how to reduce allocation overhead so the performance hit is not as large as it currently is. (And we need to do work on allocation overhead in a variety of contexts in general. E.g. the cuda thing.)

**Comment by abadams** [comment]
Note that if you want things to be promoted into registers they need to be stack allocated, so blanket turning off all stack allocations would be catastrophic. You'd probably just want to reduce the size threshold down to 32 bytes or so instead of the current several kb.

I keep thinking there should be an allocation class called a side-stack, which is just a single block of memory malloced at pipeline entry and freed at pipeline end, with statically known location of each allocation within it. This is precisely what we do for shared memory on cuda (including intelligent allocation and reuse of the space), so we could just generalize that code.

**Comment by dsharletg** [comment]
I know of at least one important pipeline in production where even a single malloc would be significant overhead (e.g. this pipeline uses `NoAsserts` and `NoBoundsQuery` because the overhead from those was significant), and one of those same pipelines is exactly one of the pipelines that suffers from stack overflows due to very small stacks. (The pipeline processes one tile of an image at a time, using the Halide FFT, which is the source of many stack overflows.)

**Comment by zvookin** [comment]
Yes, if it were easy, we'd have fixed it by now. At some level, I feel like the main goal in doing anything is to make it so Halide doesn't get blamed for the terribleness caused by the vast majority of standard C/C++ runtimes...

**Comment by dsharletg** [comment]
I'd be happy simply if the issue at runtime was reported as a stack overflow, instead of a general seg fault. I know this happens sometimes (I've seen it), but other times it seems to be reported as a more general seg fault, which is not as helpful.

Once we get that far, we now have the tool to improve this pretty easily: `store_in(MemoryType::Heap)`

**Comment by abadams** [comment]
So it would need to be scratch memory that gets passed in as an argument?
This is starting to sound like Zalman's allocation plan structure.
